File: 00-每日一题/20200516_131.cpp
class Solution {
public:
    vector<vector<string>> res;
    vector<vector<string>> partition(string s) {
        if(s.length()==0) return res;
        vector<string> temp;
        helper(temp,s,0);
        return res;
    }
    void helper(vector<string>& temp,string& s,int start){  // start表示开始字符的index
        if(s.length()==start)
        {
            res.push_back(temp);
            return;
        }
        for(int i=1;i+start-1<s.length();i++)
        {
            string x = s.substr(start,i);
            if(!ispara(x))
                continue;
            temp.push_back(x);
            helper(temp,s,start+i);
            temp.pop_back();
        }
    }

    bool ispara(string s){
        // 判断是不是回文
        for(int i=0;i<s.length()/2;i++)
        {
            if(s[i]!=s[s.length()-1-i])
                return false;
        }
        return true;
    }
};

File: 00-每日一题/20200516_131.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        res = []
        self._dfs(s, 0, [], res)
        return res

    def is_huiwen(self, s):
        i = 0
        j = len(s)-1
        while i < j:
            if s[i] != s[j]:
                return False
            i += 1
            j -= 1
        return True

    def _dfs(self, s, start, pre, res):

        if start == len(s):
            res.append(pre[:])
            return

        for i in range(start, len(s)):
            seg = s[start:i+1]
            if self.is_huiwen(seg):
                pre.append(seg)
                self._dfs(s, i+1, pre, res)
                pre.pop()


if __name__ == '__main__':
    obj = Solution()
    print(obj.is_huiwen('bbab'))
    while True:
        s = input().strip()
        res = obj.partition(s)
        print(res)